app/src/main/java/org/p2p/wallet/striga/offramp/ui/StrigaOffRampPresenter.kt

```diff
@@ -0,0 +1,344 @@
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+import org.p2p.core.network.ConnectionManager
+import org.p2p.core.utils.NoCoverage
+import org.p2p.core.utils.formatFiat
+import org.p2p.core.utils.toBigDecimalOrZero
+import org.p2p.wallet.R
+import org.p2p.wallet.common.mvp.BasePresenter
+import org.p2p.wallet.jupiter.model.SwapRateTickerState
+import org.p2p.wallet.jupiter.ui.main.mapper.SwapRateTickerMapper
+import org.p2p.wallet.striga.common.model.StrigaDataLayerError
+import org.p2p.wallet.striga.exchange.models.StrigaExchangeRate
+import org.p2p.wallet.striga.offramp.StrigaOffRampContract.Presenter
+import org.p2p.wallet.striga.offramp.StrigaOffRampContract.View
+import org.p2p.wallet.striga.offramp.interactor.StrigaOffRampInteractor
+import org.p2p.wallet.striga.offramp.mappers.StrigaOffRampMapper
+import org.p2p.wallet.striga.offramp.mappers.StrigaOffRampSwapWidgetMapper
+import org.p2p.wallet.striga.offramp.models.StrigaOffRampButtonState
+import org.p2p.wallet.striga.offramp.models.StrigaOffRampRateState
+import org.p2p.wallet.striga.offramp.models.StrigaOffRampTokenState
+import org.p2p.wallet.striga.offramp.models.StrigaOffRampTokenType
+import org.p2p.wallet.tokenservice.TokenServiceCoordinator
+import org.p2p.wallet.tokenservice.UserTokensState
+
+private const val TAG = "StrigaOffRampPresenter"
+private val DEFAULT_INITIAL_AMOUNT = BigDecimal.ZERO
+
+class StrigaOffRampPresenter(
+    private val connectionManager: ConnectionManager,
+    private val interactor: StrigaOffRampInteractor,
+    private val tokenServiceCoordinator: TokenServiceCoordinator,
+    private val strigaOffRampMapper: StrigaOffRampMapper,
+    private val swapWidgetMapper: StrigaOffRampSwapWidgetMapper,
+    private val rateTickerMapper: SwapRateTickerMapper,
+) : BasePresenter<View>(), Presenter {
+
+    private val timber: Timber.Tree = Timber.tag(TAG)
+
+    private val viewState = MutableStateFlow(
+        StrigaOffRampViewState(
+            exchangeRateState = SwapRateTickerState.Loading,
+            buttonState = StrigaOffRampButtonState.LoadingRates,
+        )
+    )
+
+    private var rate: StrigaExchangeRate? = null
+    private var balance: BigDecimal = DEFAULT_INITIAL_AMOUNT
+    private var inputAmountA: BigDecimal = DEFAULT_INITIAL_AMOUNT
+    private var inputAmountB: BigDecimal = DEFAULT_INITIAL_AMOUNT
+    private var isErrorHappened: Boolean = false
+
+    private val cannotClickAllAmount: Boolean
+        get() = isErrorHappened
+
+    override fun attach(view: View) {
+        super.attach(view)
+        interactor.startPolling(this)
+        observeUsdc()
+        observeRateChanges()
+        observeInternetState()
+        observeViewState()
+    }
+
+    override fun detach() {
+        super.detach()
+        interactor.stopPolling()
+    }
+
+    private fun observeUsdc() {
+        launch {
+            tokenServiceCoordinator.observeUserTokens()
```

Comment:
maybe we can filter here for USDC?

Reply:
it emits UserTokensState, not a List<Token.Active>, it means I would need to do the same logic as in the handleTokenState here